packages/web-client/app/components/card-pay/create-merchant-workflow/merchant-customization/index.ts

```diff
@@ -0,0 +1,151 @@
+      // no other validation errors
+      !this.merchantNameValidationMessage &&
+      !this.merchantBgColorValidationMessage
+    );
+  }
+
+  get merchantIdInputState() {
+    if (taskFor(this.validateMerchantIdTask).isRunning) {
+      return 'loading';
+    } else if (
+      this.lastCheckedMerchantId === this.merchantId &&
+      taskFor(this.validateMerchantIdTask).last?.value
+    ) {
+      return 'valid';
+    } else if (this.merchantIdValidationMessage) {
+      return 'invalid';
+    } else {
+      return 'initial';
+    }
+  }
+
+  get merchantTextColor() {
+    return mostReadable(this.merchantBgColor, [
+      '#ffffff',
+      '#000000',
+    ])!.toHexString();
+  }
+
+  @action onMerchantNameInput(value: string) {
+    this.merchantName = value;
+    this.validateMerchantName();
+  }
+
+  @action onMerchantIdInput(value: string) {
+    this.merchantId = value;
+    this.validateMerchantId();
+  }
+
+  @action onMerchantBgColorInput(event: InputEvent) {
+    let value = (event.target as HTMLInputElement).value;
+    this.merchantBgColor = value;
+  }
+
+  @action saveDetails() {
+    this.args.workflowSession.updateMany({
+      merchantName: this.merchantName,
+      merchantId: this.merchantId,
+      merchantBgColor: this.merchantBgColor,
+      merchantTextColor: this.merchantTextColor,
+    });
+    this.args.onComplete?.();
+  }
+
+  @action validateMerchantName() {
+    if (this.merchantName) {
+      this.merchantNameValidationMessage = this.merchantName.trim()
+        ? ''
+        : 'Merchant name cannot contain only spaces';
+    } else {
+      this.merchantNameValidationMessage = 'This field is required';
+    }
+  }
+
+  @action async validateMerchantId() {
+    try {
+      await taskFor(this.validateMerchantIdTask).perform();
+    } catch (e) {
+      if (didCancel(e)) {
+        return;
+      }
+      console.error(e);
+    }
+  }
+
+  @restartableTask *validateMerchantIdTask() {
+    // debounce
+    yield timeout(500);
+    let value = this.merchantId;
+
+    if (value) {
+      this.merchantIdValidationMessage = findDomainNameErrors(value);
+    } else {
+      this.merchantIdValidationMessage = 'This field is required';
+    }
+
+    if (this.merchantIdValidationMessage) {
+      return false;
+    }
+
+    // replace w call to api
+    let merchantIdExists: boolean = yield checkIfMerchantIdExists(value);
+
+    this.lastCheckedMerchantId = value;
+    if (merchantIdExists) {
+      this.merchantIdValidationMessage =
+        'This merchant ID is already taken, please choose another ID.';
+      return false;
+    }
+
+    this.merchantIdValidationMessage = '';
+    return true;
+  }
+}
+
+function findDomainNameErrors(value: string) {
+  // international domain names start with xn--
+  if (value.startsWith('xn--')) return 'Merchant ID cannot start with xn--';
+  else if (/[^a-z0-9-]/.test(value))
+    return 'Merchant ID can only contain lowercase alphabets, numbers, and hyphens';
```

Comment:
Is this string something defined elsewhere that you’re just incorporating here? I’m not used to seeing “alphabets” like this, though it’s concise at least… maybe we can ask for design/script input on the exact wording for this

Reply:
Yep, will be asking design for updated copy in a bit, including the validation/error messages.